core/types/access_witness.go

```diff
@@ -45,6 +46,113 @@ func NewAccessWitness() *AccessWitness {
 	}
 }
 
+func (aw *AccessWitness) TouchAndChargeMessageCall(addr []byte) uint64 {
+	var gas uint64
+	gas += aw.TouchAddressAndChargeGas(utils.GetTreeKeyVersion(addr[:]), nil)
+	gas += aw.TouchAddressAndChargeGas(utils.GetTreeKeyCodeSize(addr[:]), nil)
+	return gas
+}
+
+func (aw *AccessWitness) SetLeafValuesMessageCall(addr, codeSize []byte) {
+	var data [32]byte
+	aw.TouchAddress(utils.GetTreeKeyVersion(addr[:]), data[:])
+	aw.TouchAddress(utils.GetTreeKeyCodeSize(addr[:]), codeSize[:])
```

Comment:
so there's a bit of a problem with this approach: you are calling the same `GetTreeKeyVersion`/`GetTreeCodeSize` twice, once for touching it and once for setting the value. In the current implementation, that is not a problem, because this is only a hash calculation. But in the latest version of the EIP, which PR #46 addresses, it does a multi-exponentiation on an elliptic curve, and that is a performance killer. I don't think that we should care about this issue in this PR, but I would appreciate if you could put some sort of TODO here to remind us not to leave it as it.

Reply:
Added about it.